cmd/kubectl-gadget/snapshot/process.go

```diff
@@ -15,110 +15,97 @@
 package snapshot
 
 import (
-	"encoding/json"
 	"fmt"
-	"os"
 	"sort"
 	"strings"
-	"text/tabwriter"
 
 	"github.com/spf13/cobra"
 
 	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
-	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
 	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/process-collector/types"
-	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
 )
 
-var processCollectorParamThreads bool
-
-var processCollectorCmd = &cobra.Command{
-	Use:   "process",
-	Short: "Gather information about running processes",
-	RunE: func(cmd *cobra.Command, args []string) error {
-		callback := func(results []gadgetv1alpha1.Trace) error {
-			allProcesses := []types.Event{}
+type ProcessFlags struct {
+	threadGroupID bool
+}
 
-			for _, i := range results {
-				if len(i.Status.Output) == 0 {
-					continue
-				}
+type ProcessSnapshot struct {
+	outputConf   *utils.OutputConfig
+	processFlags *ProcessFlags
+}
 
-				var processes []types.Event
-				if err := json.Unmarshal([]byte(i.Status.Output), &processes); err != nil {
-					return utils.WrapInErrUnmarshalOutput(err, i.Status.Output)
-				}
+func init() {
+	processCmd := initProcessCmd()
+	SnapshotCmd.AddCommand(processCmd)
+}
 
-				allProcesses = append(allProcesses, processes...)
+func initProcessCmd() *cobra.Command {
+	var commonFlags utils.CommonFlags
+	var processFlags ProcessFlags
+
+	cmd := &cobra.Command{
+		Use:   "process",
+		Short: "Gather information about running processes",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config := &utils.TraceConfig{
+				GadgetName:       "process-collector",
+				Operation:        "collect",
+				TraceOutputMode:  "Status",
+				TraceOutputState: "Completed",
+				CommonFlags:      &commonFlags,
 			}
 
-			return printProcesses(allProcesses)
-		}
-
-		config := &utils.TraceConfig{
-			GadgetName:       "process-collector",
-			Operation:        "collect",
-			TraceOutputMode:  "Status",
-			TraceOutputState: "Completed",
-			CommonFlags:      &params,
-		}
-
-		return utils.RunTraceAndPrintStatusOutput(config, callback)
-	},
-}
+			var processGadget Snapshot[types.Event] = &ProcessSnapshot{
+				outputConf:   &commonFlags.OutputConf,
+				processFlags: &processFlags,
+			}
+			callback := getSnapshotCallback(processGadget)
 
-func init() {
-	SnapshotCmd.AddCommand(processCollectorCmd)
-	utils.AddCommonFlags(processCollectorCmd, &params)
+			return utils.RunTraceAndPrintStatusOutput(config, callback)
+		},
+	}
 
-	processCollectorCmd.PersistentFlags().BoolVarP(
-		&processCollectorParamThreads,
+	cmd.PersistentFlags().BoolVarP(
+		&processFlags.threadGroupID,
 		"threads",
 		"t",
 		false,
-		"Show all threads",
+		"Show the thread group ID",
```

Comment:
I think this doc change is confusing and the previous text ("Show all threads") was better.

When this flag is set, it will show an entry for each thread. Without this flag, only one entry per process. In multi-threaded processes, there is typically one main thread that has pid == tid.

I guess you renamed the flag to `threadGroupID` it seems to change whether the tgid is printed. But its main change is to filter out the threads (see code near `allProcessesTrimmed`).

Also, "thread group ID" is the term given in kernel code (or BPF code since it runs in the kernel). But users would normally call it "process ID" or pid:

|  Concept       | In kernel code       | In userspace code |
|---------|----------------------|-------------------|
| Process | tgid (task group id) | pid (from `getpid()`)               |
| Thread  | pid                  | tid (from `gettid()`)               |

Reply:
You are right. I got confused (again 😢 ). Thanks for the explanation!

I think my confusion comes from the name of the columns because they describe the data in "kernel terms" and not "userspace terms". Let me describe it with an example. 

When `-t` is not specified, we have the following output:

```bash
$ ./kubectl-gadget snapshot process -n gadget -p gadget-fgkll
NODE           NAMESPACE    POD             CONTAINER    COMM               PID
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559
```

It shows one process with ID `10559` (PID in terms of kernel and userspace). So far, so good. 

Now, if I want to know if that process has more than one thread, I use the `-t` option:

```bash
$ ./kubectl-gadget snapshot process -n gadget -p gadget-fgkll -t
NODE           NAMESPACE    POD             CONTAINER    COMM               TGID     PID
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10559
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10586
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10587
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10588
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10589
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10590
vm-minikube    gadget       gadget-fgkll    gadget       gadgettracerman    10559    10591
...
```

Here is where I don't think the name of the columns describes very well the situation in terms of userspace. I mean, I read it as the column PID now represents the thread ID and the TGID, which is kernel terminology, represents the PID. Don't you think?

IMO, after thinking about it better, the description should remain "Show all threads" as you said. Then, if our idea is to show the information in terms of userspace, I think we should change the name of the last two columns to be PID and TID (thread ID):

```bash
$ ./kubectl-gadget snapshot process -n gadget -p gadget-fgkll -t
NODE           NAMESPACE    POD             CONTAINER    COMM               PID     TID
```

What do you think?

BTW, I found [this](https://stackoverflow.com/a/9306150) comment very useful.

Reply:
I agree, the columns should use the userspace terminology

Reply:
I agree too, but this can be dealt in another PR.

Reply:
I drop that change from this PR, and I will change the columns header in another PR.